File: backend/app/schemas/resume.py
import datetime
import uuid
from typing import Any, Literal
from pydantic import BaseModel, ConfigDict, Field, field_validator


class SkillAnalysis(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    category: str | None = None
    proficiency: str | None = None

    @field_validator("name")
    @classmethod
    def strip_name(cls, v: str) -> str:
        return v.strip()


class ExperienceAnalysis(BaseModel):
    company: str | None = None
    job_title: str | None = None
    location: str | None = None
    employment_type: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    description: str | None = None
    is_current: bool = False


class ProjectAnalysis(BaseModel):
    project_name: str | None = None
    description: str | None = None
    technologies: list[str] = Field(default_factory=list)
    project_url: str | None = None
    start_date: str | None = None
    end_date: str | None = None


class EducationAnalysis(BaseModel):
    institution: str | None = None
    degree: str | None = None
    field_of_study: str | None = None
    start_date: str | None = None
    end_date: str | None = None
    grade: str | None = None


class CertificationAnalysis(BaseModel):
    name: str | None = None
    issuing_organization: str | None = None
    issue_date: str | None = None
    expiration_date: str | None = None
    credential_url: str | None = None


class ResumeAnalysis(BaseModel):
    full_name: str | None = None
    email: str | None = None
    phone: str | None = None
    location: str | None = None
    linkedin_url: str | None = None
    github_url: str | None = None
    portfolio_url: str | None = None
    professional_summary: str | None = None
    target_roles: list[str] = Field(default_factory=list)
    years_of_experience: float | None = None
    skills: list[SkillAnalysis | str] = Field(default_factory=list)
    experience: list[ExperienceAnalysis] = Field(default_factory=list)
    projects: list[ProjectAnalysis] = Field(default_factory=list)
    education: list[EducationAnalysis] = Field(default_factory=list)
    certifications: list[CertificationAnalysis] = Field(default_factory=list)

    @field_validator("email", mode="before")
    @classmethod
    def clean_email(cls, v: Any) -> Any:
        if isinstance(v, str):
            v = v.strip().lower()
            v = v.rstrip(".") if v.endswith(".") else v
        return v or None


class ResumeUploadResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    filename: str = Field(validation_alias="original_filename")
    file_type: str
    file_size: int
    version: int
    is_active: bool
    status: Literal["uploaded"] = "uploaded"
    profile_created: bool = False


class ResumeListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    filename: str = Field(validation_alias="original_filename")
    file_type: str
    file_size: int
    version: int
    is_active: bool
    has_profile: bool = False
    created_at: datetime.datetime


class ResumeDetail(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    filename: str = Field(validation_alias="original_filename")
    file_type: str
    file_size: int
    version: int
    is_active: bool
    created_at: datetime.datetime

File: frontend/types/resume.ts
export interface ResumeItem {
  id: string;
  filename: string;
  file_type: string;
  file_size: number;
  version: number;
  is_active: boolean;
  has_profile: boolean;
  created_at: string;
}

export interface ResumeUploadResponse {
  id: string;
  filename: string;
  file_type: string;
  file_size: number;
  version: number;
  is_active: boolean;
  status: "uploaded";
  profile_created: boolean;
}

export interface Skill {
  id: string;
  skill_name: string;
  skill_category: string;
  proficiency?: string | null;
}

export interface Experience {
  id: string;
  company: string;
  job_title: string;
  location?: string | null;
  employment_type?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  description?: string | null;
  is_current?: boolean;
}

export interface Project {
  id: string;
  project_name: string;
  description?: string | null;
  technologies: string[];
  project_url?: string | null;
  start_date?: string | null;
  end_date?: string | null;
}

export interface Education {
  id: string;
  institution: string;
  degree?: string | null;
  field_of_study?: string | null;
  start_date?: string | null;
  end_date?: string | null;
  grade?: string | null;
}

export interface Certification {
  id: string;
  name: string;
  issuing_organization?: string | null;
  issue_date?: string | null;
  expiration_date?: string | null;
  credential_url?: string | null;
}

export interface CandidateProfile {
  id: string;
  resume_id?: string | null;
  full_name?: string | null;
  email?: string | null;
  phone?: string | null;
  location?: string | null;
  linkedin_url?: string | null;
  github_url?: string | null;
  portfolio_url?: string | null;
  professional_summary?: string | null;
  target_roles: string[];
  years_of_experience?: number | null;
  skills: Skill[];
  experience: Experience[];
  projects: Project[];
  education: Education[];
  certifications: Certification[];
  updated_at: string;
}

export interface Completeness {
  percentage: number;
  missing: string[];
}